src/arch/xen/include/xen_memory.h

```diff
@@ -0,0 +1,127 @@
+/**
+ * @file    xen_memory.h
+ * @brief
+ *
+ * @author  Aleksei Cherepanov
+ * @date    30.03.2020
+ */
+
+#ifndef XEN_MEMORY_
+#define XEN_MEMORY_
+
+#include <unistd.h>
+#include <xen/xen.h>
+
+#if 1
+//phymem reuse
+#include <mem/page.h>
+extern struct page_allocator *xen_mem_allocator;
+
+#define __xen_mem_end (__xen_mem_allocator->pages_start + \
+		(__xen_mem_allocator->page_size * __xen_mem_allocator->pages_n))
+
+extern void *xen_mem_alloc(size_t page_number);
+extern void xen_mem_free(void *page, size_t page_number);
+
+extern struct page_allocator *xen_allocator(void);
+// end of phymem
+#endif
+
+extern unsigned long *phys_to_machine_mapping;
+extern start_info_t xen_start_info;
+extern shared_info_t xen_shared_info;
+
+#define pfn_to_mfn(_pfn) (phys_to_machine_mapping[(_pfn)])
+#define mfn_to_pfn(_mfn) (machine_to_phys_mapping[(_mfn)])
+
+
+#define L1_PAGETABLE_SHIFT      12
+#define L2_PAGETABLE_SHIFT      21
+#define L3_PAGETABLE_SHIFT      30
+
+#define L1_PAGETABLE_ENTRIES    512
+#define L2_PAGETABLE_ENTRIES    512
+#define L3_PAGETABLE_ENTRIES    4
+
+/* Given a virtual address, get an entry offset into a page table. */
+#define l1_table_offset(_a) \
+  (((_a) >> L1_PAGETABLE_SHIFT) & (L1_PAGETABLE_ENTRIES - 1))
+#define l2_table_offset(_a) \
+  (((_a) >> L2_PAGETABLE_SHIFT) & (L2_PAGETABLE_ENTRIES - 1))
+#define l3_table_offset(_a) \
+  (((_a) >> L3_PAGETABLE_SHIFT) & (L3_PAGETABLE_ENTRIES - 1))
+
+#define VIRT_START                 ((unsigned long)0x100000L) //hardcode of ELFNOTE(XEN_ELFNOTE_VIRT_BASE, .long 0x100000)
+#define to_phys(x)                 ((unsigned long)(x)-VIRT_START)
+#define PFN_DOWN(x)	((x) >> L1_PAGETABLE_SHIFT)
+#define virt_to_pfn(_virt)         (PFN_DOWN(to_phys(_virt)))
+
+#define virt_to_mfn(_virt)         (pfn_to_mfn(virt_to_pfn(_virt)))
+
+#define CONST(x) x ## ULL
+#define _PAGE_PRESENT  CONST(0x001)
```

Comment:
xen_memory.h should be a public interface only:
* alloc/free
* probably virt2mfn

Declarations/macros that are used only in xen_memory.c are likely to be implementation details, not interface. Implementation details should go to .c file.